File: plugins/ghidra/src/main/java/lief/dwarf/editor/CompilationUnit.java
/* Copyright 2022 - 2026 R. Thomas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package lief.dwarf.editor;

public class CompilationUnit extends lief.Base {
    @Override
    protected native void destroy();

    private CompilationUnit(long impl) {
        super(impl);
    }

    public native void setProducer(String producer);

    public native Function createFunction(String name);

    public native Variable createVariable(String name);

    public native Type createGenericType(String name);

    public native EnumType createEnum(String name);

    public native TypeDef createTypedef(String name, Type type);

    public native StructType createStructure(String name, StructType.Type kind);

    public StructType createStructure(String name) {
        return createStructure(name, StructType.Type.STRUCT);
    }

    public native BaseType
        createBaseType(String name, int size, BaseType.Encoding encoding);

    public BaseType createBaseType(String name, int size) {
        return createBaseType(name, size, BaseType.Encoding.NONE);
    }

    public native FunctionType createFunctionType(String name);

    public PointerType createPointerType(Type ty) {
        return ty.getPointerTo();
    }

    public native Type createVoidType();

    public native ArrayType createArray(String name, Type ty, int count);
};

File: api/rust/crates/lief-ffi/src/dwarf/editor/compilation_unit.rs
#[cxx::bridge]
pub mod ffi {
    unsafe extern "C++" {
        include!("LIEF/rust/DWARF/editor/CompilationUnit.hpp");

        type DWARF_editor_ArrayType = crate::dwarf::editor::array_type::ffi::DWARF_editor_ArrayType;
        type DWARF_editor_BaseType = crate::dwarf::editor::base_type::ffi::DWARF_editor_BaseType;
        type DWARF_editor_EnumType = crate::dwarf::editor::enum_type::ffi::DWARF_editor_EnumType;
        type DWARF_editor_Function = crate::dwarf::editor::function::ffi::DWARF_editor_Function;
        type DWARF_editor_FunctionType =
            crate::dwarf::editor::function_type::ffi::DWARF_editor_FunctionType;
        type DWARF_editor_PointerType =
            crate::dwarf::editor::pointer_type::ffi::DWARF_editor_PointerType;
        type DWARF_editor_StructType =
            crate::dwarf::editor::struct_type::ffi::DWARF_editor_StructType;
        type DWARF_editor_Type = crate::dwarf::editor::type_::ffi::DWARF_editor_Type;
        type DWARF_editor_TypeDef = crate::dwarf::editor::type_def::ffi::DWARF_editor_TypeDef;
        type DWARF_editor_Variable = crate::dwarf::editor::variable::ffi::DWARF_editor_Variable;

        type DWARF_editor_CompilationUnit;

        fn set_producer(self: Pin<&mut DWARF_editor_CompilationUnit>, value: &CxxString);
        fn create_function(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
        ) -> UniquePtr<DWARF_editor_Function>;
        fn create_variable(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
        ) -> UniquePtr<DWARF_editor_Variable>;
        fn create_generic_type(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
        ) -> UniquePtr<DWARF_editor_Type>;
        fn create_enum(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
        ) -> UniquePtr<DWARF_editor_EnumType>;
        fn create_typedef(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
            ty: &DWARF_editor_Type,
        ) -> UniquePtr<DWARF_editor_TypeDef>;
        fn create_structure(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
            kind: u32,
        ) -> UniquePtr<DWARF_editor_StructType>;
        fn create_base_type(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
            size: u64,
            encoding: u32,
        ) -> UniquePtr<DWARF_editor_BaseType>;
        fn create_function_type(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
        ) -> UniquePtr<DWARF_editor_FunctionType>;
        fn create_pointer_type(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            ty: &DWARF_editor_Type,
        ) -> UniquePtr<DWARF_editor_PointerType>;
        fn create_void_type(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
        ) -> UniquePtr<DWARF_editor_Type>;
        fn create_array_type(
            self: Pin<&mut DWARF_editor_CompilationUnit>,
            name: &CxxString,
            ty: &DWARF_editor_Type,
            count: u64,
        ) -> UniquePtr<DWARF_editor_ArrayType>;
    }

    impl UniquePtr<DWARF_editor_CompilationUnit> {}
}
